src/uct/ib/dc/dc_mlx5_ep.c

```diff
@@ -906,6 +907,21 @@ ucs_status_t uct_dc_mlx5_ep_fc_pure_grant_send(uct_dc_mlx5_ep_t *ep,
     UCT_DC_MLX5_CHECK_DCI_RES(iface, ep);
     UCT_DC_MLX5_IFACE_TXQP_GET(iface, ep, txqp, txwq);
 
+    if (iface->flags & UCT_DC_MLX5_IFACE_FLAG_FC_EP_FAILED) {
+        /* If the EP has FAILED flag set, it means that the DCI has already been
+         * assigned to the FC endpoint and it has not been reset yet. It is
+         * currently waiting for all operation compeltions as a part of handling
+         * failure. Return NO_RESOURCE status to add the operation to the
+         * pending queue for furhter processing after error recovery. FC EP
+         * arbiter group mustn't be scheduled to avoid dispatching pending
+         * operations while DCI isn't recovered from the failure */
+        ucs_assertv(dci != UCT_DC_MLX5_EP_NO_DCI, "fc_ep %p: dci %u", ep, dci);
```

Comment:
this assert can be removed since it's after UCT_DC_MLX5_CHECK_DCI_RES

Reply:
`dci` is assigned before `UCT_DC_MLX5_CHECK_DCI_RES`

Reply:
> `dci` is assigned before `UCT_DC_MLX5_CHECK_DCI_RES`

yes, so, we have to check this to make sure this is correct error flow

Reply:
is it possible we passed UCT_DC_MLX5_CHECK_DCI_RES check without exiting, and still have no DCI?

Reply:
> is it possible we passed UCT_DC_MLX5_CHECK_DCI_RES check without exiting, and still have no DCI?

but I'm checking the previous DCI allocated, it should be allocated in case of failure

Reply:
seems excessive to me to check previous ep->dci state

Reply:
> seems excessive to me to check previous ep->dci state

why not have it? it is need to check that everything is correct with error flow

Reply:
because there is no part in error flow that relies on fc_ep->dci being valid when fc_ep is failed, adding too many non-essential assertions will make it harder to make changes later on

Reply:
> because there is no part in error flow that relies on fc_ep->dci being valid when fc_ep is failed

actually, we rely on it here:
https://github.com/openucx/ucx/blob/0b60458cb526a4399f8016960bcec998ed155f8c/src/uct/ib/dc/dc_mlx5_ep.c#L1556
the endpoint should own a DCI until it has outstanding operations. when no outstanding operation, the DCI must be released and reset and FAILED flag should be removed from the EP. So, here we rely on the fact that FAILED flag is set when DCI is assigned on our EP.

> adding too many non-essential assertions will make it harder to make changes later on

yes, I agree with you. But here we just make sure that we don't break anything with error flow for FC EP.

Reply:
fixed